cudax/include/cuda/experimental/__stf/utility/cuda_attributes.cuh

```diff
@@ -0,0 +1,29 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDASTF in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if defined(__CUDACC__) && !defined(_NVHPC_CUDA)
+#  define CUDASTF_NO_DEVICE_STACK
+#elif defined(__CUDACC__) && defined(_NVHPC_CUDA)
+#  define CUDASTF_NO_DEVICE_STACK _Pragma("diag_suppress no_device_stack")
+#else
+#  define CUDASTF_NO_DEVICE_STACK
+#endif
```

Comment:
This could be 
```suggestion
#if defined(_CCCL_CUDA_COMPILER_NVHPC)
#  define CUDASTF_NO_DEVICE_STACK _Pragma("diag_suppress no_device_stack")
#else // ^^^ _CCCL_CUDA_COMPILER_NVHPC ^^^ / vvv !_CCCL_CUDA_COMPILER_NVHPC vvv
#  define CUDASTF_NO_DEVICE_STACK
#endif // !_CCCL_CUDA_COMPILER_NVHPC
```

@dkolsen-pgi Is that meant to annotate a lambda because we are currently never setting that warnings suppression back which concerns me

Reply:
it was a mean to help the compiler when we had a lambda calling a lambda on the device, probably an extended lambda calling a lambda ...

Reply:
Yeah, but we are using it inconsistently.

there are 2 occurences. one adds it to both lambdas but none disable that warning again, so I am wondering whether this acts like a "warning" or an local attribute.

If it is the former we need to reestablish the warnin after we use it

Reply:
The uses are:

```cpp
    // Help the compiler which may not detect that a device lambda is calling a device lambda
    CUDASTF_NO_DEVICE_STACK
    return make_tuple_indexwise<dimensions>([&](auto i) {
      // included
      const ::std::ptrdiff_t begin_i  = get_begin(i);
      const ::std::ptrdiff_t extent_i = get_extent(i);
      auto result            = begin_i + (index % extent_i);
      index /= extent_i;
      return result;
    });
    CUDASTF_NO_DEVICE_STACK
```
```cpp
  CUDASTF_NO_DEVICE_STACK
  auto explode_args = [&](auto... data) {
    // For every linearized index in the shape
    for (; i < n; i += step)
    {
      CUDASTF_NO_DEVICE_STACK
      auto explode_coords = [&](auto... coords) {
        f(coords..., data...);
      };
      ::std::apply(explode_coords, shape.index_to_coords(i));
    }
  };
  ::std::apply(explode_args, mv(targs));
```

Reply:
> probably an extended lambda calling a lambda

There are no extended lambdas with NVC++.  A lambda is just a lambda and it just works on both host and device, even generic lambdas with captures.  Except when a lambda that captures variables by reference is created in host code and then called from device code, where the reference captures will be illegal memory references.  That's what the warning is for, since it is an easy mistake for a user to make when using stdpar.

When a lambda that captures by reference is created in a device-only function and is only ever called from device code, the reference captures are not a problem and the compiler doesn't warn about that.

But there are some situation involving nested lambdas where the reference captures are fine (device code will never try to access anything on the host), but NVC++ can't figure that out and emits a false warning.  I tried really hard to get the compiler to not warn in these situations, but the compiler doesn't have the right information at the right time to do the right thing.

So suppressing the warning in some situations is necessary.  It should be rare.  And Michael is correct that the warning should be turned back on after the lambda.